daemon/exec.go

```diff
@@ -293,15 +296,22 @@ func (daemon *Daemon) ContainerExecStart(ctx context.Context, name string, optio
 	select {
 	case <-ctx.Done():
 		logrus.Debugf("Sending TERM signal to process %v in container %v", name, c.ID)
-		daemon.containerd.SignalProcess(ctx, c.ID, name, signal.SignalMap["TERM"])
+		sigCtx, cancelFunc := context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancelFunc()
+		daemon.containerd.SignalProcess(sigCtx, c.ID, name, signal.SignalMap["TERM"])
 
-		timeout := time.NewTimer(termProcessTimeout)
+		wait := termProcessTimeout
+		if ec.StopTimeout != nil {
+			wait = *ec.StopTimeout
+		}
```

Comment:
If we're making the timeout configurable anyway, what do you think of dropping the arbitrary default value and leave it entirely up to the caller to dictate the policy? I'm thinking along the lines of [mirroring this proposal for `os/exec`](https://github.com/golang/go/issues/50436) which defaults to insta-kill on cancel while allowing for both the graceful-terminate signal and timeout to be configured.

Reply:
To clarify, I am not suggesting we expose any features to users. My line of thinking is that the API handler `func (*containerRouter) postContainerExecCreate` would be a more appropriate place to dictate the stop-timeout policy for execs created via the API, even if it is the existing hardcoded `termProcessTimeout` policy, just like how `cmdProbe` now dictates the policy for health-check execs. The obvious implementation would require adding a `StopTimeout` field to `types.ExecConfig`, though tagging it <tt>\`json:"-"\`</tt> would keep it from being exposed in the Engine API, just like its analogue in `container.HealthConfig`.

Reply:
Not 100% sure if those defaults should make its way into the API endpoint, but perhaps in `Daemon.ContainerExecCreate()`, as that's where the exec config is created (and to match what I did for the health check) https://github.com/moby/moby/blob/4eb1c5bd52afa08595d9a082a9ba05c81f79be07/daemon/exec.go#L98

That way the code here (`ContainerExecStart`) wouldn't have to bother with checking if all options are there, and just consume the exec-config as stored. Perhaps make `exec.NewConfig()` accept options so that we don't have the awkward code in both locations to first do a `exec.NewConfig()`, and then having to manually manipulate it to set all the other options.

Given that the health check code doesn't use the endpoint, nor `Daemon.ContainerExecCreate()`, I think that would also be fine to do separately though 🤔 WDYT?

Reply:
I totally agree default values should be handled while initializing struct, vs checking for nil to apply defaults.
Constructors is one of the language feature I miss the most writing Go. I'd be happy we rely on `NewXX` initializer pattern everywhere possible

Reply:
> I totally agree default values should be handled while initializing struct, vs checking for nil to apply defaults.

I should note that it's not always possible to do so when constructing things; some default should be "baked" into a struct, whereas some other defaults may have to be evaluated at runtime (e.g. if option X isn't set, take the current default as configured in the environment and/or daemon configuration).

But I agree that having a canonical location to where defaults are set (_where possible_) is good. Currently there's too many places where defaults are set, which means you often have to drill through a lot of code to figure out where they're set.

Reply:
^^ we should not increase the scope of this PR too much though; I haven't checked yet what would be needed, but I'd still be fine to tackle that in a follow-up (more "improve exec in general, not just for health-check").

Perhaps just moving this to `ContainerExecCreate()` would work though;

```go
if ec.StopTimeout != nil {
	wait = *ec.StopTimeout
}
```

Which could be just a line added here; https://github.com/moby/moby/blob/4eb1c5bd52afa08595d9a082a9ba05c81f79be07/daemon/exec.go#L128


```go
execConfig.StopTimeout = termProcessTimeout
```